### PR TITLE
Add train_stream verb for streaming training on live data sources

### PR DESCRIPTION
**Pull request by drewoldag**
## Change Description

Adds a new `train_stream` verb that enables training models on open-ended streaming data sources (e.g., Kafka topics) using a context-manager session pattern. This mirrors the existing `infer_stream` verb but for training workflows.

### New Components

**`TrainStream` verb** (`src/hyrax/verbs/train_stream.py`):
- Loads a model once and returns a `TrainStreamSession` for on-demand batch training
- Supports two modes:
  1. **Data-source driven**: Configure `[data_request.train_stream]` with a streaming dataset; the session is iterable and yields `(batch, metrics)` pairs
  2. **Manual**: Pass a representative `sample_batch` and feed batches via `session.train_batch(batch)`
- Reuses existing infrastructure: `create_process_func("train_batch", ...)` for the per-batch training step, `setup_model` / `setup_model_from_sample` for model initialization, and MLflow + TensorBoard logging

**`TrainStreamSession` context manager**:
- Iterates a streaming data loader or accepts manual batches
- Skips empty batches and optionally enforces `min_batch_size` (guards against size-1 batches breaking BatchNorm in user models)
- Logs per-batch metrics to TensorBoard and MLflow
- Periodically checkpoints weights (every N batches if `save_weights_every` is set)
- Persists final weights and ends MLflow run on `close()`
- Returns the trained model

### Configuration

Added `[train_stream]` section to `hyrax_default_config.toml`:
- `model_weights_file`: warm-start weights path (default: `false` = train from scratch)
- `weights_filename`: checkpoint filename (default: `"example_model.pth"`)
- `save_weights_every`: checkpoint interval in batches (default: `false` = only on close)
- `experiment_name`: MLflow experiment (default: `"notebook"`)
- `run_name`: MLflow run name (default: `false` = use results-dir name)
- `min_batch_size`: skip batches smaller than this (default: `false` = train all)

### Design Rationale

- **No epoch loop**: Streams are open-ended, so we drive a per-batch loop directly instead of using Ignite's `create_trainer` / `trainer.run(..., max_epochs=...)` path
- **Partial batches**: Streaming sources yield ragged batches (flushed on timeout); we train every non-empty batch as-is, with optional `min_batch_size` guard
- **Gradient weighting**: Each batch steps the optimizer independently (no accumulation); accepted for streaming workflows
- **Code sharing**: Mirrors `infer_stream` pattern; no shared-base refactor to avoid over-engineering

### Testing

Added comprehensive test suite (`tests/hyrax/test_train_stream.py`):
- Data-source-driven iteration with `KafkaStreamDataset` (using `FakeConsumer` stand-in)
- Manual session with spy `process_func` to verify empty/small-batch skipping
- Idempotent `close()` and error handling
- CLI raises `NotImplementedError` (programmatic API only)

## Code Quality
- [x] I have read the Contribution Guide and agree to the Code of Conduct
- [x] My code follows the code style of this project
- [x] My code builds cleanly without errors or warnings
- [x] My code contains relevant comments and necessary documentation

https://claude.ai/code/session_01Vhimu4YPAyMBPj3dPzEmp4

### CODE REVIEW

**Comment by drewoldag** [comment]
Note for Drew for later to run the hyrax dataset producer to emit HSC images:
`python hyrax_dataset_producer.py --dataset hsc --data-dir <base_path>/hsc_8asec_1000 --burst-min=75 --burst-max=100 --num-bursts=100 --delay-min=1 --delay-max=5`

File: src/hyrax/verbs/train_stream.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+
+        # Skip empty / too-small batches. Streaming sources emit ragged batches (a partial
+        # batch is flushed when the latency timeout fires), and a size-1 batch can break
+        # batch-size-sensitive layers (e.g. BatchNorm) in some user models.
+        num_samples = self._batch_num_samples(batch)
+        if num_samples is not None:
+            if num_samples == 0:
+                logger.debug("Skipping empty batch.")
+                return None
+            min_batch_size = self._config["train_stream"]["min_batch_size"]
```
Comment (drewoldag) [inline src/hyrax/verbs/train_stream.py:301]: Make sure this is build up the batch and not drop it on the floor.

File: src/hyrax/verbs/train_stream.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+
+        if mlflow.active_run() is not None:
+            mlflow.end_run()
+
+        self._close_logger()
+        logger.info("TrainStream session closed.")
+        return self._model
+
+    def checkpoint(self, model_metrics=None) -> None:
+        """Save the current model weights if loss is lower than the previous best.
```
Comment (drewoldag) [inline src/hyrax/verbs/train_stream.py:373]: Double check on this to see if it's still used.

File: src/hyrax/datasets/lsdb_stream_dataset.py [hunk 3]
```diff
@@ -0,0 +1,989 @@
...
+
+        import lsdb
+
+        open_catalog_kwargs = dict(ds_config["open_catalog_kwargs"] or {})
+        if requested_columns and "columns" not in open_catalog_kwargs:
+            open_catalog_kwargs["columns"] = requested_columns
+
+        return lsdb.open_catalog(self.data_location, **open_catalog_kwargs)
+
+    def _requested_columns_from_config(self, config: dict):
```
Comment (drewoldag) [inline src/hyrax/datasets/lsdb_stream_dataset.py:575]: Can we just get rid of this entirely? We should just load all the columns.

Comment (drewoldag) [reply src/hyrax/datasets/lsdb_stream_dataset.py:575]: This functionality is in HATSDataset, so we could just call it directly.

File: src/hyrax/datasets/lsdb_stream_dataset.py [hunk 4]
```diff
@@ -0,0 +1,989 @@
...
+            # host/port semantics and mangle any name containing a colon.
+            return self._lookup_catalog(self.data_location[len(LSDB_URI_PREFIX) :])
+
+        import lsdb
+
+        open_catalog_kwargs = dict(ds_config["open_catalog_kwargs"] or {})
+        if requested_columns and "columns" not in open_catalog_kwargs:
+            open_catalog_kwargs["columns"] = requested_columns
+
+        return lsdb.open_catalog(self.data_location, **open_catalog_kwargs)
```
Comment (drewoldag) [inline src/hyrax/datasets/lsdb_stream_dataset.py:573]: I would argue for removing this functionality. We should just expect users to hand Hyrax the pandas future. We should then be able to remove `_requested_columns_from_config`

File: src/hyrax/datasets/lsdb_stream_dataset.py [hunk 5]
```diff
@@ -0,0 +1,989 @@
...
+                "config['data_set']['LSDBStreamDataset']['prefetch_chunks'] to overlap the "
+                "fetches with a background thread instead."
+            )
+            return None
+
+        logger.info(f"The lsdb stream is using the active dask client: {client}")
+        self._warn_on_small_chunks(client)
+        return client
+
+    def _warn_on_small_chunks(self, client) -> None:
```
Comment (drewoldag) [inline src/hyrax/datasets/lsdb_stream_dataset.py:674]: Maybe don't do this warning thing. This is copied from the LSDB documentation. We don't need to warn users about it. This is not our fight, this is LSDB and Dask.

File: src/hyrax/datasets/lsdb_stream_dataset.py [hunk 6]
```diff
@@ -0,0 +1,989 @@
...
+        batch: list[dict] = list(self._buffered)
+        self._buffered = []
+        self._peek_index = 0
+
+        # The chunk currently being drained, and how far into it we have read. Tracked out
+        # here so the `finally` can push back the rows of a half-drained chunk.
+        rows: list[dict] = []
+        taken = 0
+
+        try:
```
Comment (drewoldag) [inline src/hyrax/datasets/lsdb_stream_dataset.py:903]: Make sure that left over rows are held until the next chunk, and then combined with the first batch.

File: src/hyrax/datasets/lsdb_stream_dataset.py [hunk 7]
```diff
@@ -0,0 +1,989 @@
...
+            # fetched - are pushed back so a later pass resumes exactly where this one
+            # stopped, in order and without duplicates.
+            leftover = rows[taken:] + self._release_source()
+            if batch or leftover:
+                self._buffered = batch + leftover + self._buffered
+            if self._exhausted or self._stop.is_set():
+                self._iterator = None
+            self._timings.log()
+
+    def collate_lightcurve(self, samples):
```
Comment (drewoldag) [inline src/hyrax/datasets/lsdb_stream_dataset.py:965]: This should be tossed out. It was for experimentation.

**Comment by drewoldag** [comment]
@maxwest-uw - @mtauraso and I came to the conclusion that the best path forward for this PR is likely to leave it as a draft, create a fresh branch and pull over only the important pieces. There has been a lot of experimentation along the way and as a result also a lot of superfluous code, notebooks, etc.

Feel free to take a look over it and comment if you have time though :)
